src/pybind/ceph_argparse.py

```diff
@@ -1317,9 +1317,8 @@ def run_in_thread(func, *args, **kwargs):
     if timeout == 0 or timeout == None:
         # python threading module will just get blocked if timeout is `None`,
         # otherwise it will keep polling until timeout or thread stops.
-        # wait for INT32_MAX, as python 3.6.8 use int32_t to present the
-        # timeout in integer when converting it to nanoseconds
-        timeout = (1 << (32 - 1)) - 1
+        # wait for 5 seconds, as python 3.6.8 use float to represent the timeout
```

Comment:
strictly speaking, both float and integer are accepted in python 3.6.8. see https://github.com/python/cpython/blob/v3.6.8/Python/pytime.c#L315 .

the reason why i put `(1 << (32 - 1)) - 1` here is to explain that we don't want to encode a magic number here, and user wants to wait forever if timeout is not specified.

any specific reason you want to use a smaller number here, aside from using "float" to follow the document?

Reply:
On 32 bit system I get pretty much immediate "time out" exception here. I guess it happens because of some sort of overflow in python when doing bit shift. I will check what python puts in timeout = (1 << (32 - 1)) - 1 on 32 bit machine.

Reply:
2^31 is just over 2 seconds. Correct? I tried to put 5.0 to effectively double that - on slow machine timeout is hit just because it is slow, not because something went badly wrong.

Reply:
thanks. i double checked the CPython source. it's using `int64_t` to represent the `ns` helded by a `PyTime`. see https://github.com/python/cpython/blob/master/Include/pytime.h#L19 . so that part should be safe even after converting from seconds to nanosecond by multiplying it with 10^9.

Reply:
2 ^ 31 is 2147483648 seconds. see https://github.com/python/cpython/blob/v3.6.8/Python/pytime.c#L338

Reply:
@vbashkirtsev how about replacing it with `threading.TIMEOUT_MAX` ? which is much better than both magic numbers.

Reply:
In python on 32 bit system:

timeout = (1 << (32 - 1)) - 1 = 2147483647

It is 2.147 seconds and is way too low on slow systems. Hence my 5.0 passing. But I agree that magic numbers are bad idea.

Your intention with (1 << (32 - 1)) - 1 was "user wants to wait forever if timeout is not specified"? I am a bit confused.

Reply:
> In python on 32 bit system:
> 
> timeout = (1 << (32 - 1)) - 1 = 2147483647
> 
> It is 2.147 seconds and is way too low on slow systems. Hence my 5.0 passing. But I agree that magic numbers are bad idea.

could reconsider my reply at https://github.com/ceph/ceph/pull/38272#discussion_r530118269.  see https://github.com/python/cpython/blob/3c6b436a57893dd1fae4e072768f41a199076252/Python/pytime.c#L338-L342

or am i missing something obvious?

> 
> Your intention with (1 << (32 - 1)) - 1 was "user wants to wait forever if timeout is not specified"? I am a bit confused.

in this very python function, if the caller does not pass `timeout`, it implies that it wants to wait forever. and apparently, "forever" is not 5 second.

Reply:
Apologies. It is me who missing obvious :) Yes, you're right - int provides seconds, not nanoseconds as I was thinking. So 2^31 should be much more than 5.0!

I have tried threading.TIMEOUT_MAX and it fails with immediate "timed out" exception as well. Checked its value and it is 9223372036.0 which coming from /usr/include/limits.h LLONG_MAX divided by 1000000000. _PyTime_t is 64 bit int - plenty of room to fit.

I am totally dumbfounded by this bug - will debug it deeper as it seems the issue is somewhere in Python. It is just not acceptable to blindly stick in 5.0 without understanding why the hell it fails!

Reply:
I probably going to speculate: I did look inside of Python to see why join() returns before timeout. It appears that it depends on implementation of monotonic clock on the system. On armv7l clock apparently operates differently to one on x86 or even on aarch64. In python's thread_pthread.h there some calculations like "timeout + _PyTime_GetMonotonicClock()" which on armv7l apparently overflow and it returns negative value which in effect means "time is up" and thus it bails out with "timed out". If I reduce timeout in ceph_argparse.py to (1 << 30 - 1)) - 1 then "magically" everything works because sum of timeout and monotonic clock value does not overflow. So, despite that I have not found exact point of overflow, all indications are that it is overflow somewhere in Python or kernel and it is clearly beyond our control (not that we cannot patch Python or kernel but just because it is out of scope of ceph code).

So I came back to ceph_argparse.py to see what else can we do. For join() timeout may be 0 - do not block, timeout may be number of seconds (supposedly up to threading.TIMEOUT_MAX but now we know it is buggy) - block for no more than timeout period, timeout may be 'None' - block indefinitely. If intention is "user wants to wait forever if timeout is not specified" should not we just do the following?
```
if timeout == 0:
    timeout = None
```
kwargs.pop('timeout', 0) returns 0, not 'None' if no timeout is set and for join() it means "do not block" while intention is to block indefinitely - so we just need to change timeout to 'None' if it is 0. I tested with above code and it works on armv7l. No magic numbers involved and works as advertised. What do you think?

Reply:
@vbashkirtsev thanks for looking into the root cause . but i cannot find in `timeout + _PyTime_GetMonotonicClock()` https://github.com/python/cpython/blob/v3.6.8/Python/thread_pthread.h . it exists in the master branch though.

so what overflows is the product of `timeout + _PyTime_GetMonotonicClock()`. because `deadline` would be so large as it is a `_PyTime_t` which is an alias of `long long`. in other words, `threading.TIMEOUT_MAX` is not a safe number of `Lock.acquire()`, even https://github.com/python/cpython/blob/master/Python/thread_pthread.h#L440 ensures that `microseconds` does not overflow after being multiplied with 10^3. because
```c
deadline = _PyTime_GetMonotonicClock() + timeout;
```
could  also overflow if the sum of these two values exceeds `LLONG_MAX`.

i cannot recall the details. but we don't want the python thread to be **blocked**. see the comment in the code. i am quoting it here

> python threading module will just get blocked if timeout is `None`,

how about using "1 day" as an synonym of "infinite" or "forever", so we are safe until the 292,277,026,596 AD - 1 day. i.e., using following patch
```diff
diff --git a/src/pybind/ceph_argparse.py b/src/pybind/ceph_argparse.py
index ac985e5113e..3e6a8b95fb5 100644
--- a/src/pybind/ceph_argparse.py
+++ b/src/pybind/ceph_argparse.py
@@ -1318,8 +1318,11 @@ def run_in_thread(func, *args, **kwargs):
         # python threading module will just get blocked if timeout is `None`,
         # otherwise it will keep polling until timeout or thread stops.
         # wait for INT32_MAX, as python 3.6.8 use int32_t to present the
-        # timeout in integer when converting it to nanoseconds
-        timeout = (1 << (32 - 1)) - 1
+        # timeout in integer when converting it to nanoseconds, but since
+        # python3 uses `int64_t` for the deadline before timeout expires,
+        # we have to use a safe value which does not overflow after being
+        # added to current time in microseconds.
+        timeout = 24 * 60 * 60
     t = RadosThread(func, *args, **kwargs)

     # allow the main thread to exit (presumably, avoid a join() on this
```

what do you think?

Reply:
Tried 24 * 60 * 60 on armv7l and tests do pass. So I think it is the best solution we should adopt.

Reply:
un-resolving this thread. as the suggested change is not applied.